### ISSUE TITLE
Document reactor.listenMulticast's listenMultiple keyword argument

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4504]
|<img alt="aalex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| aalex reported|
|-|-|
|Trac ID|trac#4504|
|Type|enhancement|
|Created|2010-06-14 22:11:00Z|
|Branch|https://github.com/twisted/twisted/tree/listenmulticast-docs-4504|

@param listenMultiple: If set to True, allows multiple servers to listen to the same address and port of a multicast group.
@type listenMultiple: bool

Some commonly used addresses are the All Hosts multicast group, such ass 224.0.0.1, that contains all systems on the same network segment. 

<details><summary>Searchable metadata</summary>

```
trac-id__4504 4504
type__enhancement enhancement
reporter__aalex aalex
priority__normal normal
milestone__ 
branch__branches_listenmulticast_docs_4504 branches/listenmulticast-docs-4504
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation_easy documentation easy
time__1276553460000000 1276553460000000
changetime__1279030333000000 1279030333000000
version__None None
owner__ 
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_1"></a>|@jesstess set owner to aalex|
|-|-|

aalex is going to give a go at a patch today! ==> Reassigning.

**Comment by twisted-trac** [comment]
|<img alt="aalex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|aalex commented|
|-|-|

Here is my patch:

```
Index: udp.py
===================================================================
--- udp.py	(revision 29319)
+++ udp.py	(working copy)
@@ -309,6 +309,12 @@
     implements(interfaces.IMulticastTransport)
 
     def __init__(self, port, proto, interface='', maxPacketSize=8192, reactor=None, listenMultiple=False):
+        """
+        @type listenMultiple: bool
+        @param listenMultiple: If set to True, allows multiple servers to listen to the same address and port of a multicast group.
+
+        Some commonly used multicast addresses are the All Hosts multicast group, such ass 224.0.0.1, that contains all systems on the same network segment.
+        """
         Port.__init__(self, port, proto, interface, maxPacketSize, reactor)
         self.listenMultiple = listenMultiple
```

**Comment by twisted-trac** [comment]
|<img alt="aalex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|aalex commented|
|-|-|

I updated my comment a bit:
```
Index: twisted/internet/udp.py
===================================================================
--- twisted/internet/udp.py	(revision 29319)
+++ twisted/internet/udp.py	(working copy)
@@ -309,6 +309,14 @@
     implements(interfaces.IMulticastTransport)
 
     def __init__(self, port, proto, interface='', maxPacketSize=8192, reactor=None, listenMultiple=False):
+        """
+        @type listenMultiple: C{bool}
+        @param listenMultiple: If set to True, allows multiple servers to listen to the same address and port of a multicast group.
+
+        Some commonly used multicast addresses are the All Hosts multicast group, such ass 224.0.0.1, that contains all systems on the same network segment.
+        
+        To join a multicast group, it might be useful to create a subclass of t.p.DatagramProtocol in order to override its startProtocol method. In it, one can call self.transport.joinGroup(224.0.0.1) to join the desired multicast group, which is 224.0.0.1 in this example.
+        """
         Port.__init__(self, port, proto, interface, maxPacketSize, reactor)
         self.listenMultiple = listenMultiple
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess set owner to aalex|
|-|-|

Thanks for working on this, aalex! Some feedback:

 * Please line wrap at 79 columns. More on Twisted's coding conventions can be found [here](http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
 * I would leave out the example and let people reference the RFC if they'd like to learn more about multicast.
 * Since this docstring is already getting added, can the rest of the parameters get markup too? I'd say "steal from Port", but Port doesn't document its init parameters either (if you're up for both, even better!).

Please attach subsequent diffs instead of pasting into a comment. :)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess commented|
|-|-|

I'm also happy to do the parts of the feedback you don't have time for. Thanks again for the patch!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Also, some of this documentation looks like it would be better off in doc/core/howto/udp.xhtml.  epytext markup for parameters usually follows the descriptive part of a docstring, rather than the other way around.  And don't forget a news fragment. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29629]) Branching to 'listenmulticast-docs-4504'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29630]) Apply aalex's patch

refs [#4504](https://github.com/twisted/twisted/issues/4504)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

I changed things around a bit.  There _is_ api doc coverage for `listenMultiple` already.  I'm not sure why aalex was looking in twisted/internet/udp.py instead of at the interface.  I put a `@see` in to try to help out the matter.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/listenmulticast-docs-4504)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set owner to @exarkun|
|-|-|

"all hosts group" is a bit mystifying, how about "the special address for all hosts on the local network segment (see http://www.faqs.org/rfcs/rfc1112.html)" if that is accurate? We probably should have a reference to the RFC in the howto regardless.

Other than that (and need for news file) looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [29800]) Merge listenmulticast-docs-4504

Author: aalex, exarkun
Reviewer: itamar
Fixes: [#4504](https://github.com/twisted/twisted/issues/4504)

Improve the discoverability of information about multicast, in particular
the `listenMultiple` argument to `IReactorMulticast.listenMulticast`.  Also
drop a reference to the RFC into the UDP howto.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
